File: 33-Search-in-Rotated-Sorted-Array/solution.java
public class Solution {
    public int search(int[] A, int target) {
        //rotated array can be seperated into two segment
        
        if(A.length == 0 || A == null) return -1;
        
        int start = 0;
        int end = A.length - 1;
        while(start <= end) {
            int mid = start + (end - start)/2;
            if(target == A[mid]) return mid;

            if(A[mid] >= A[start]){
                if(A[start] <= target && target <= A[mid]){
                    end = mid - 1;
                }else{
                    start = mid + 1;
                }
            }else{
                if (A[mid] <= target && target <= A[end]) {
                    start = mid + 1;
                }else{
                    end = mid - 1;
                }

            }
        }
        
        return -1;
    }
}

File: 33-Search-in-Rotated-Sorted-Array/solution.py
class Solution(object):
    def search(self, nums, target):

        s = 0
        e = len(nums)-1

        while s<=e:
            mid = (s+e)/2

            vmid = nums[mid]
            vs = nums[s]
            ve = nums[e]

            if vmid == target:
                return mid
            elif vs == target:
                return s
            elif ve == target:
                return e

            if ((target-vs) * (target - vmid) <= 0 and vs < vmid) or (vmid < ve and (target-ve) * (target - vmid) > 0):
                e = mid-1
            else:
                s = mid+1
        return -1